libs/ml/include/ml/ops/weights.hpp

```diff
@@ -57,6 +59,36 @@ class Trainable
   virtual ArrayType const &get_gradients() const                                     = 0;
   virtual void             ResetGradients()                                          = 0;
   virtual void             ApplyGradient(ArrayType const &grad)                      = 0;
+
+  void SetRegularisation(fetch::ml::details::RegularisationType regularisation,
+                         DataType                               regularisation_rate = DataType{0.0})
+  {
+    this->regularisation_      = regularisation;
+    this->regularisation_rate_ = regularisation_rate;
+  }
+
+  void ApplyRegularisation()
+  {
+    switch (this->regularisation_)
+    {
+    case fetch::ml::details::RegularisationType::L1:
+      L1Regularisation(this->regularisation_rate_);
+      break;
+    case fetch::ml::details::RegularisationType::L2:
+      L2Regularisation(this->regularisation_rate_);
+      break;
+    default:
+      break;
+    }
+  }
+
+protected:
+  fetch::ml::details::RegularisationType regularisation_ =
+      fetch::ml::details::RegularisationType::NONE;
+  DataType regularisation_rate_ = static_cast<DataType>(0);
+
+  virtual void L1Regularisation(DataType regularisation_rate) = 0;
```

Comment:
I think it would be better if you make an abstract class called Regulariser. then make child classes L1Regularisation and L2Regularisation. Then weights.hpp can have a regulariser_ private member variable; Then SetRegularisation constructs the variable, and you don't need the switch case in apply regularisation anymore - you just invoke the regulariser.

This way, the regularisation implementations can live in ml/regularisers rather than being implemented directly in weights.